lnwallet/rebroadcaster.go

```diff
@@ -0,0 +1,335 @@
+package lnwallet
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/btcsuite/btcd/blockchain"
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil"
+)
+
+var (
+	// ErrRebroadcastMaxAttempts is an error returned when the rebroadcaster
+	// has reached its maximum number of allowed unsuccessful attempts.
+	ErrRebroadcastMaxAttempts = errors.New("exhausted maximum number of " +
+		"rebroadcast attempts")
+
+	// ErrRebroadcastMaxFee is an error returned when the rebroadcaster has
+	// reached the maximum fee allowed after bumping up the fee due to many
+	// unsuccessful attempts
+	ErrRebroadcastMaxFee = errors.New("exceeded rebroadcast max fee")
+
+	// ErrRebroadcastDustLimit is an error returned when the rebroadcaster
+	// is applying a new fee to the transaction but would result in the
+	// change output dipping below the dust limit.
+	ErrRebroadcastDustLimit = errors.New("change output dipped below " +
+		"dust limit")
+
+	// ErrRebroadcastHalt is an error returned when the rebroadcaster has
+	// been requested to halt.
+	ErrRebroadcastHalt = errors.New("rebroadcaster is exiting")
+)
+
+// WitnessSignDescriptor is a tuple of a transaction input's sign descriptor
+// along with its witness type. This is needed in order to use the generic
+// GenWitnessFunc in order to generate the correct witness for each type.
+type WitnessSignDescriptor struct {
+	WitnessType
+	*SignDescriptor
+}
+
+// TxRebroadcasterCfg contains all of the required fields to be able to bump a
+// transaction's current fee, re-sign it, and broadcast it to the network.
+//
+// NOTE: All fields are required unless otherwise noted.
+type TxRebroadcasterCfg struct {
+	// Tx is the transaction that should be rebroadcast.
+	Tx *wire.MsgTx
+
+	// ChangeOutputIndex is the index of the change output that should be
+	// used to bump the transaction's fee.
+	ChangeOutputIndex uint32
```

Comment:
As rebroadcaster is rebuilding txes anyway, wouldn't it be better to pass the data in in a little more flexible format, not yet a `MsgTx`?

Reply:
What type for format?

Reply:
I was thinking a list of inputs with indices, the change output separately and a list of the other non-change outputs. And also bundle the sign descriptors with the inputs. Like at precursor for the to be generated `MsgTx` that contains just the right data fields.